File: Go/Splitwise-go/balance_sheet.go
package main

import "fmt"

type BalanceSheet struct {
	balances []*Balance
}

func NewBalanceSheet() *BalanceSheet {
	return &BalanceSheet{balances: make([]*Balance, 0)}
}

func (b *BalanceSheet) AddDebt(debtorID, creditorID int, amount int64) error {
	if debtorID == creditorID {
		return nil
	}
	if amount <= 0 {
		return fmt.Errorf("amount must be a positive integer (paise)")
	}

	var existing *Balance
	for _, bal := range b.balances {
		if bal.DebtorID == debtorID && bal.CreditorID == creditorID {
			existing = bal
			break
		}
	}

	var opposite *Balance
	oppIdx := -1
	for i, bal := range b.balances {
		if bal.DebtorID == creditorID && bal.CreditorID == debtorID {
			opposite = bal
			oppIdx = i
			break
		}
	}

	if opposite != nil && opposite.Amount > 0 {
		if opposite.Amount > amount {
			opposite.Amount -= amount
			return nil
		}
		amount -= opposite.Amount
		b.balances = append(b.balances[:oppIdx], b.balances[oppIdx+1:]...)
	}

	if existing != nil {
		existing.Amount += amount
		return nil
	}
	if amount <= 0 {
		return nil
	}
	b.balances = append(b.balances, &Balance{
		DebtorID:   debtorID,
		CreditorID: creditorID,
		Amount:     amount,
	})
	return nil
}

func (b *BalanceSheet) GetPairwiseBalances() []*Balance {
	return b.balances
}

func (b *BalanceSheet) GetBalance(debtorID, creditorID int) *Balance {
	for _, bal := range b.balances {
		if bal.DebtorID == debtorID && bal.CreditorID == creditorID {
			return bal
		}
	}
	return nil
}

File: JavaScript/Splitwise/BalanceSheet.js
const { Balance } = require('./Balance');

class BalanceSheet {
  constructor() {
    this.balances = [];
  }

  addDebt(debtorId, creditorId, amount) {
    if (debtorId === creditorId) return;
    if (!Number.isInteger(amount) || amount <= 0) {
      throw new Error('amount must be a positive integer (paise)');
    }

    const existingBalance = this.balances.find(balance => balance.debtorId === debtorId && balance.creditorId === creditorId);

    const opposite = this.balances.find(balance => balance.debtorId === creditorId && balance.creditorId === debtorId);
    if (opposite && opposite.amount > 0) {
      if (opposite.amount > amount) {
        opposite.amount -= amount;
        return;
      }

      // Remove the opposite balance as it is fully paid
      this.balances.splice(this.balances.indexOf(opposite), 1);
      amount -= opposite.amount;
    }

    if (existingBalance) {
      existingBalance.amount += amount;
      return;
    }

    if (amount <= 0) {
      return;
    }

    this.balances.push(new Balance(debtorId, creditorId, amount));
  }

  getPairwiseBalances() {
    return this.balances;
  }

  getBalance(debtorId, creditorId) {
    return this.balances.find(balance => balance.debtorId === debtorId && balance.creditorId === creditorId);
  }
}

module.exports = { BalanceSheet };
